File: VB/Reporting_how-to-add-a-pre-created-parameter-instance-to-the-field-list-and-keep-the-e4247/Form2.vb
Imports System
Imports System.ComponentModel
Imports System.ComponentModel.Design
Imports System.Windows.Forms
Imports DevExpress.XtraReports.Design
Imports DevExpress.XtraReports.Parameters

Namespace Parameters
    Partial Public Class Form2
        Inherits Form

        Public Sub New()
            InitializeComponent()
        End Sub

        Private Sub barButtonItem1_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs) Handles barButtonItem1.ItemClick
            Dim parameter As New Parameter()
            parameter.Name = "MyDateTimeParameter"
            parameter.Type = GetType(Date)
            xrDesignPanel1.Report.Parameters.Add(parameter)
            AddField(parameter, "MyParameter")
        End Sub

        Private Sub AddField(ByVal objeto As IComponent, ByVal desiredName As String)
            Dim host As IDesignerHost = TryCast(xrDesignPanel1.GetService(GetType(IDesignerHost)), IDesignerHost)
            If host IsNot Nothing Then
                DesignTool.AddToContainer(host, objeto, desiredName)
            End If
        End Sub
    End Class
End Namespace

File: CS/Reporting_how-to-add-a-pre-created-parameter-instance-to-the-field-list-and-keep-the-e4247/Form2.cs
using System;
using System.ComponentModel;
using System.ComponentModel.Design;
using System.Windows.Forms;
using DevExpress.XtraReports.Design;
using DevExpress.XtraReports.Parameters;

namespace Parameters {
    public partial class Form2 : Form {
        public Form2() {
            InitializeComponent();
        }

        private void barButtonItem1_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e) {
            Parameter parameter = new Parameter();
            parameter.Name = "MyDateTimeParameter";
            parameter.Type = typeof(DateTime);
            xrDesignPanel1.Report.Parameters.Add(parameter);
            AddField(parameter, "MyParameter");
        }

        private void AddField(IComponent objeto, string desiredName) {
            IDesignerHost host = xrDesignPanel1.GetService(typeof(IDesignerHost)) as IDesignerHost;
            if (host != null) {
#pragma warning disable CS0612 // Type or member is obsolete
                DesignToolHelper.AddToContainer(host, objeto, desiredName);
#pragma warning restore CS0612 // Type or member is obsolete
            }
        }
    }
}
